### PR TITLE
implement simple OrcJIT adaptor for Julia

### PR DESCRIPTION
**Pull request by vtjnash**
this provides a toy implementation of the (eager) llvm OrcJIT interface for experimenting and testing

### CODE REVIEW

**Comment by Keno** [comment]
Cool! Lang Hames and I were planning to sit down on Wed to discuss what else we need in ORC jit to make it work really well in julia. Now we'll have something to start from. Also with the move to ORC jit, I was planning to move back to having julia emit into a single module and let ORC jit handle the splitting. Is that how this works?

File: src/codegen.cpp [hunk 1]
```diff
@@ -1053,6 +1221,19 @@ void *jl_function_ptr(jl_function_t *f, jl_value_t *rt, jl_value_t *argt)
     Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt);
     assert(llvmf);
     JL_GC_POP();
+
+#ifdef USE_MCJIT
```
Comment (Keno) [inline src/codegen.cpp]: Why is this needed now, and shouldn't the statement just below it be deleted?

Comment (Keno) [inline src/codegen.cpp]: Or did you mean this to be for ORC jit?

Comment (vtjnash) [inline src/codegen.cpp]: i just this from below. this is a more logical place for it

Comment (Keno) [inline src/codegen.cpp]: Ok, I guess I was confused why there because there's another `ifdef USE_MCJIT` right after.

**Comment by vtjnash** [comment]
> Also with the move to ORC jit, I was planning to move back to having julia emit into a single module and let ORC jit handle the splitting. Is that how this works?

agreed. right now this acts more like the eager MCJIT, since that was simpler

File: src/codegen.cpp (hunk 1 above)
Comment (vtjnash) [inline src/codegen.cpp]: eh, yeah. probably could have merged those

**Comment by JeffBezanson** [comment]
Very cool!

File: src/codegen.cpp [hunk 2]
```diff
@@ -5753,6 +5930,20 @@ static void init_julia_llvm_env(Module *m)
...
+    jlpowf_func = Function::Create(FunctionType::get(T_float32, powf_type, false),
+                                   Function::ExternalLinkage,
+                                   "powf", m);
+    add_named_global(jlpowf_func, (void*)&powf);
+
+    Type *pow_type[2] = { T_float64, T_float64 };
+    jlpow_func = Function::Create(FunctionType::get(T_float64, pow_type, false),
+                                  Function::ExternalLinkage,
+                                  "pow", m);
+    add_named_global(jlpow_func, (void*)&pow);
```
Comment (tkelman) [inline src/codegen.cpp]:
```
d:/code/msys64/home/Tony/julia/src/codegen.cpp(5944) : error C2440: 'type cast'
: cannot convert from 'overloaded-function' to 'void *'
        Context does not allow for disambiguation of overloaded function
d:/code/msys64/home/Tony/julia/src/codegen.cpp(5944) : error C2660: 'add_named_global' : function does not take 1 arguments
```

Comment (vtjnash) [inline src/codegen.cpp]: sigh msvc, that's not quite the function I wanted, but the C++ `<cmath>` header is shadowing the desired `pow` function from `math.h`. i think we might need to use dlsym as an escape hatch, but i'm open to other ideas.

Comment (yuyichao) [inline src/codegen.cpp]: Would `::pow` work?

Comment (tkelman) [inline src/codegen.cpp]: do you mean like
```diff
                                   "pow", m);
-    add_named_global(jlpow_func, (void*)&pow);
+    add_named_global(jlpow_func, (void*)&(::pow));
 #endif
```
? apparently not

Comment (yuyichao) [inline src/codegen.cpp]: Yeah, I just thought maybe it's confusing `::pow` with `std::pow` (which is [the one that is overloaded](http://en.cppreference.com/w/cpp/numeric/math/pow)). I guess MSVC is probably doing sth else then....

In that case does `static_cast<float (*)(float, float)>(&pow)` (with or without the `&`) help? ....

Comment (yuyichao) [inline src/codegen.cpp]: Sorry I mean `static_cast<double (*)(double, double)>(&pow)`

Comment (tkelman) [inline src/codegen.cpp]: Looks like that'll work. It compiles, anyway.

Comment (vtjnash) [inline src/codegen.cpp]: that'll manage to call the "untrusted" system libm, rather than the statically linked openlibm

@yuyichao from SO, it sounds to me like that's how it was supposed to be implemented originally, but some more recent C++ standard has changed that to acknowledge that some compiler implementations just reused the C header + overloads.

**Comment by yuyichao** [comment]
Actually the binary is broken right after `make`. I thought it worked because I was doing the wrong test. (My local copy outside the packaging directory is too old and running `julia -f -e 0` actually don't use any function in the sysimg....)

c.c. @Keno mainly because I'm rebasing your branch on the current master to get through bootstrap and test this...

**Comment by yuyichao** [comment]
Printing the function name in `emit_function` suggested that the missing function is the one that is being emitted.

```
emit_function, julia_pointer_22378
emit_function, julia_convert_22379
emit_function, julia_start_22380
emit_function, julia_set_print_level_22381
emit_function, julia_convert_22382
emit_function, julia_finalizer_22383
emit_function, julia_get_22384
emit_function, julia_unsafe_convert_22385
               _
   _       _ _(_)_     |  A fresh approach to technical computing
  (_)     | (_) (_)    |  Documentation: http://docs.julialang.org
   _ _   _| |_  __ _   |  Type "?help" for help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 0.5.0-dev+1017 (2015-10-29 14:40 UTC)
 _/ |\__'_|_|_|\__'_|  |  kf/improvedi/16c31af (fork: 1 commits, 0 days)
|__/                   |  x86_64-unknown-linux-gnu

emit_function, julia_anonymous_22386
emit_function, julia_repl_color_22387
emit_function, julia_anonymous_22388
emit_function, julia_call_22389
emit_function, julia___Prompt#184___22390
emit_function, julia_call_22391
LLVM ERROR: Program used external function 'julia_call_22391' which could not be resolved!
```

**Comment by yuyichao** [comment]
OK. so the minimum sub-patch that I could come up with that reproduce this issue is

```diff
 src/codegen.cpp | 45 +++++++++++++++++++++++++++------------------
 1 file changed, 27 insertions(+), 18 deletions(-)

diff --git a/src/codegen.cpp b/src/codegen.cpp
index f0a3d23..dbd5ec2 100644
--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -862,13 +862,10 @@ static void jl_setup_module(Module *m, bool add)
     if (jl_ExecutionEngine)
         m->setDataLayout(jl_ExecutionEngine->getDataLayout());
 #endif
-    if (add) {
-        assert(jl_ExecutionEngine);
-#ifdef LLVM36
-        jl_ExecutionEngine->addModule(std::unique_ptr<Module>(m));
-#else
-        jl_ExecutionEngine->addModule(m);
-#endif
+}
+
+static void jl_finalize_module(Module *m)
+{
 #if defined(_CPU_X86_64_) && defined(_OS_WINDOWS_) && defined(USE_MCJIT)
         ArrayType *atype = ArrayType::get(T_uint32,3); // want 4-byte alignment of 12-bytes of data
         (new GlobalVariable(*m, atype,
@@ -878,7 +875,12 @@ static void jl_setup_module(Module *m, bool add)
             false, GlobalVariable::InternalLinkage,
             ConstantAggregateZero::get(atype), "__catchjmp"))->setSection(".text");
 #endif
-    }
+    assert(jl_ExecutionEngine);
+#if defined(LLVM36) && !defined(USE_ORCJIT)
+    jl_ExecutionEngine->addModule(std::unique_ptr<Module>(m));
+#else
+    jl_ExecutionEngine->addModule(m);
+#endif
 }
 
 extern "C" void jl_generate_fptr(jl_function_t *f)
@@ -908,6 +910,7 @@ extern "C" void jl_generate_fptr(jl_function_t *f)
         #endif
 
         Function *llvmf = (Function*)li->functionObject;
+        jl_finalize_module(llvmf->getParent());
 #ifdef USE_MCJIT
         li->fptr = (jl_fptr_t)(intptr_t)jl_ExecutionEngine->getFunctionAddress(llvmf->getName());
 #else
@@ -1053,6 +1056,19 @@ void *jl_function_ptr(jl_function_t *f, jl_value_t *rt, jl_value_t *argt)
     Function *llvmf = jl_cfunction_object(f, rt, (jl_tupletype_t*)argt);
     assert(llvmf);
     JL_GC_POP();
+
+#ifdef USE_MCJIT
+    if (uint64_t addr = jl_ExecutionEngine->getFunctionAddress(llvmf->getName()))
+        return (void*)(intptr_t)addr;
+    if (llvmf->getParent() == shadow_module) {
+        // Copy the function out of the shadow module
+        Module *m = new Module("julia", jl_LLVMContext);
+        jl_setup_module(m, true);
+        FunctionMover mover(m, shadow_module);
+        (void)mover.CloneFunction(llvmf);
+        jl_finalize_module(m);
+    }
+#endif
 #ifdef USE_MCJIT
     return (void*)(intptr_t)jl_ExecutionEngine->getFunctionAddress(llvmf->getName());
 #else
@@ -4042,16 +4058,9 @@ static Function *gen_cfun_wrapper(jl_function_t *ff, jl_value_t *jlrettype, jl_t
 #endif
 
 #ifdef USE_MCJIT
-    if (imaging_mode) {
-        // Copy the function out of the shadow module
-        Module *m = new Module("julia", jl_LLVMContext);
-        jl_setup_module(m, true);
-        FunctionMover mover(m, shadow_module);
-        Function *clone = mover.CloneFunction(cw);
-        FPM->run(*clone);
-    }
-    else {
-        FPM->run(*cw);
+    FPM->run(*cw);
+    if (!imaging_mode) {
+        jl_finalize_module(m);
     }
 #endif
```

(The patch has some redudent parameters and indentation inconsistency in order to minimize the noise in it). I checked each `jl_setup_module(..., true)` and find that the one in `emit_function` doesn't have a corresponding `jl_finalize_module`, while all other ones does. Since AFAICT, part of the code in `jl_setup_module` to register the module that was conditioned on the `add` parameter is not moved to `jl_finalized_module`, this is probably why I see this problem.

The following patch fixes my issue, whic I'll submit in a PR soon. I'm not 100% sure if this is the right solution due to my limitted understanding of this issue.
```diff
diff --git a/src/codegen.cpp b/src/codegen.cpp
index dbd5ec2..49b6e8d 100644
--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -766,6 +766,8 @@ void jl_dump_compiles(void *s)
     dump_compiles_stream = (JL_STREAM*)s;
 }
 
+static void jl_finalize_module(Module *m);
+
 // --- entry point ---
 //static int n_emit=0;
 static Function *emit_function(jl_lambda_info_t *lam);
@@ -815,6 +817,11 @@ static Function *to_function(jl_lambda_info_t *li)
     }
 #endif
     FPM->run(*f);
+#ifdef USE_MCJIT
+    if (!imaging_mode) {
+        jl_finalize_module(f->getParent());
+    }
+#endif
     //n_compile++;
     // print out the function's LLVM code
     //jl_static_show(JL_STDERR, (jl_value_t*)li);
```
